core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -138,12 +142,55 @@ class SnappySessionState(val snappySession: SnappySession)
     Seq(ConditionalPreWriteCheck(datasources.PreWriteCheck(conf, catalog)), PrePutCheck)
   }
 
-  override lazy val analyzer: Analyzer = new Analyzer(catalog, conf) {
+  private lazy val snappyAnalyzer: Analyzer = new Analyzer(catalog, conf) {
 
     override val extendedResolutionRules: Seq[Rule[LogicalPlan]] =
       getExtendedResolutionRules(this)
 
+  }
+
+  override lazy val analyzer = new Analyzer(catalog, conf) {
+
     override val extendedCheckRules: Seq[LogicalPlan => Unit] = getExtendedCheckRules
+
+    private def getStrategy(strategy: snappyAnalyzer.Strategy): Strategy = strategy match {
+      case snappyAnalyzer.FixedPoint(_) => fixedPoint
+      case _ => Once
+    }
+
+    override lazy val batches: Seq[Batch] = snappyAnalyzer.batches.map {
+      case batch  if batch.name.equalsIgnoreCase("Resolution") =>
+        val rules = batch.rules.flatMap {
+          // Prepending PreUpdateTypeConversionCheck rule before PromoteString rule
+          case PromoteStrings =>
+            Seq(StringPromotionCheckForUpdate.asInstanceOf[Rule[LogicalPlan]],
+              PromoteStrings)
+          case r => Seq(r)
+        }
+
+        Batch(batch.name, getStrategy(batch.strategy), rules: _*)
+      case batch => Batch(batch.name, getStrategy(batch.strategy), batch.rules: _*)
+    }
+  }
+
+  // This Rule fails an update query when type of Arithmetic operators doesn't match. This
+  // need to be done because by default spark performs null safe implicit type
+  // conversion when type of two operands does't match and this can lead to null values getting
+  // populated in the table.
+  object StringPromotionCheckForUpdate extends Rule[LogicalPlan] {
+
+    override def apply(plan: LogicalPlan): LogicalPlan = {
+      plan match {
+        case Update(_, _, _, _, _) => plan resolveExpressions {
+          case e if !e.childrenResolved => e
+          case BinaryArithmetic(_@StringType(), _) | BinaryArithmetic(_, _@StringType()) =>
+            throw new AnalysisException("Implicit type casting is not performed for update" +
+                " statements")
+          case e => e
+        }
+        case _ => plan
+      }
+    }
```

Comment:
I think we discussed that this should be applied only for updateExpressions, not in other places for consistency right?